File: leet-code/641. Design Circular Deque/DesignCircularDeque641_3.java
package com.example;

public class DesignCircularDeque641_3 {

    private final int maxSize;
    private int size = 0;
    Node head;
    Node tail;

    public DesignCircularDeque641_3(int k) {
        maxSize = k;
        head = new Node(-1);
        tail = new Node(-1);
        head.next = tail;
        tail.prev = head;
    }

    public boolean insertFront(int value) {
        if (isFull()) {
            return false;
        }

        Node node = new Node(value);
        Node oldHead = head.next;
        oldHead.prev = node;
        head.next = node;
        node.prev = head;
        node.next = oldHead;

        size += 1;
        return true;
    }

    public boolean insertLast(int value) {
        if (isFull()) {
            return false;
        }

        Node node = new Node(value);
        Node oldTail = tail.prev;
        oldTail.next = node;
        tail.prev = node;
        node.prev = oldTail;
        node.next = tail;

        size += 1;
        return true;
    }

    private void insertWhenIsEmpty(int value) {
        head = new Node(value);
        tail = head;
    }

    public boolean deleteFront() {
        if (isEmpty()) {
            return false;
        }

        Node headNext = head.next.next;
        Node deleteNode = head.next;
        head.next = headNext;
        headNext.prev = head;
        deleteNode.prev = null;
        deleteNode.next = null;

        size -= 1;
        return true;
    }

    public boolean deleteLast() {
        if (isEmpty()) {
            return false;
        }

        Node tailPrev = tail.prev.prev;
        Node deleteNode = tail.prev;
        tail.prev = tailPrev;
        tailPrev.next = tail;
        deleteNode.prev = null;
        deleteNode.next = null;

        size -= 1;
        return true;
    }

    private void deleteWhenIsSizeOne() {
        head = null;
        tail = null;
    }

    public int getFront() {
        if (isEmpty()) {
            return -1;
        }
        return head.next.value;
    }

    public int getRear() {
        if (isEmpty()) {
            return -1;
        }
        return tail.prev.value;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return maxSize == size;
    }

    private static class Node {
        int value;
        Node prev;
        Node next;

        Node(int value) {
            this.value = value;
        }
    }
}

File: leet-code/622. Design Cirular Queue/DesignCircularDeque641_3.kt
package com.example

class DesignCircularDeque641_3(k: Int) {
    var head: Node? = null
    var tail: Node? = null
    var curSize = 0
    val maxSize = k

    fun insertFront(value: Int): Boolean {
        val onInsert: (Node) -> Unit = { node ->
            val oldHead = head
            head = node
            head!!.next = oldHead
            oldHead!!.prev = head
        }
        return insert(value, onInsert)
    }

    fun insertLast(value: Int): Boolean {
        val onInsert: (Node) -> Unit = { node ->
            val oldTail = tail
            tail = node
            tail!!.prev = oldTail
            oldTail!!.next = tail
        }
        return insert(value, onInsert)
    }

    private fun insert(value: Int, onInsert: (Node) -> Unit): Boolean {
        if (isFull()) return false
        val node = Node(value)
        if (isEmpty()) {
            head = node
            tail = node
        } else {
            onInsert(node)
        }
        curSize += 1
        return true
    }

    fun deleteFront(): Boolean {
        val onDelete: () -> Unit = {
            val next = head!!.next
            head = next
            head!!.prev = null
        }
        return delete(onDelete)
    }

    fun deleteLast(): Boolean {
        val onDelete: () -> Unit = {
            val prev = tail!!.prev
            tail = prev
            tail!!.next = null
        }
        return delete(onDelete)
    }

    private fun delete(onDelete: () -> Unit): Boolean {
        if (isEmpty()) return false
        if (curSize == 1) {
            head = null
            tail = null
        } else {
            onDelete()
        }
        curSize -= 1
        return true
    }

    fun getFront(): Int {
        return if (isEmpty()) -1 else head!!.value
    }

    fun getRear(): Int {
        return if (isEmpty()) -1 else tail!!.value
    }

    fun isEmpty(): Boolean {
        return curSize == 0
    }

    fun isFull(): Boolean {
        return curSize == maxSize
    }

    class Node(k: Int) {
        var prev: Node? = null
        var next: Node? = null
        val value = k
    }
}
